dynamax/linear_gaussian_ssm/inference.py

```diff
@@ -215,9 +249,26 @@ def _condition_on(m, P, H, D, d, R, u, y):
          mu_pred (D_hid,): predicted mean.
          Sigma_pred (D_hid,D_hid): predicted covariance.
     """
-    # Compute the Kalman gain
-    S = R + H @ P @ H.T
-    K = psd_solve(S, H @ P).T
+    if R.ndim == 2:
+        S = R + H @ P @ H.T
+        K = psd_solve(S, H @ P).T
+    else: 
+        # Optimization using Woodbury identity with A=R, U=H@chol(P), V=U.T, C=I
+        # (see https://en.wikipedia.org/wiki/Woodbury_matrix_identity)
+
+        R_inv = jnp.diag(1.0 / R)
```

Comment:
Rather than instantiating a diagonal matrix and then doing matrix-matrix multiplies below (line 262), you can just do a matrix-vector multiplies to further speed up the computation. The quantity you need is $R^{-1/2} U$. With that, I think you can simplify this to,
```
X = U / jnp.sqrt(R)[:, None]
S_inv = jnp.diag(1.0 / R) - X @ psd_solve(I + X.T @ X, X.T)

Reply:
Good idea! The outer R's shouldn't be square-rooted so I put
```
X = U / R[:, None]
S_inv = jnp.diag(1.0/R) - X @ psd_solve(I + U.T @ X, X.T) 
```